PWA first tutorial: first 3 parts (#26891)
mdn/content: files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md

        <p>
          <label for="end-date">End date</label>
          <input type="date" id="end-date" required />
        </p>
      </fieldset>
      <p>
        <button type="submit">Add Period</button>
      </p>
    </form>
    <section id="past-periods"></section>
    <script src="app.js" defer></script>
  </body>
</html>
```

If you haven't already created the file, copy this HTML and save it in a file called `index.html`.

Note the last line in the body links to the `app.js` script. This is the script we are creating in this section. In this lesson, we will be writing client-side JavaScript to capture form submissions, locally store submitted data, and populate the past periods section.

At the end of this lesson, you will have a fully functional app. In future lessons, we will progressively enhance the app to create an fully installable PWA that works even when the user is offline.

## Javascript task

When a user visits the page, we check if they have existing data stored in local storage. The first time a user visits the page, there won't be any data. When a new user selects two dates and submits the form, we need to:

1. Create a "`<h2>Past periods</h2>`" header

> 2. Create an {{HTMLelement("ul")}}

Author: i debated doing this. It was easier to add a line of JS then to explain the UX design requirements of empty data along with the CSS for the two different experiences. (the "empty" message would not be a list item semantically)

  // Reset the form.
  newPeriodFormEl.reset();
});
```

After preventing the form submission with [`preventDefault()`](/en-US/docs/Web/API/Event/preventDefault), we:

1. [Validate user input](#validate-user-input); exiting if invalid,
2. store the new cycle by [retrieving, parsing, appending, sorting, stringifying, and re-storing](#retrieve-append-sort-and-re-store-data) data in localStorage,
3. [render the form data](#render-data-to-screen) along with the data of past menstrual cycles and a section header, and
4. reset the form using the HTMLFormElement [`reset()`](/en-US/docs/Web/API/HTMLFormElement/reset) method

### Validate user input

We check if the dates are invalid. We do minimal error checking. We make sure neither date is null, which the `required` attribute should prevent from happening. We also check that the start date is not greater than the end date. If there is an error, we clear the form.

```javascript
function checkDatesInvalid(startDate, endDate) {
  // Check that end date is after start date and neither is null.
  if (!startDate || !endDate || startDate > endDate) {
    // For now, we clear the dates if either
    // or both are invalid

>     newPeriodFormEl.reset();

Author: I understand and agree with what you are saying. If we were creating a robust app, this would not reset the form, but rather do the steps explained in the comments. That is why this was created as a separate function -- for the comments, not the current code. I reordered the comments in the function to make this clearer.

  console.log(periods);
  return periods;
}

function renderPastPeriods() {
  const pastPeriodHeader = document.createElement("h2");
  const pastPeriodList = document.createElement("ul");
  const periods = getAllStoredPeriods();
  if (periods.length === 0) {
    return;
  }
  pastPeriodContainer.innerHTML = "";
  pastPeriodHeader.textContent = "Past periods";
  periods.forEach((period) => {
    const periodEl = document.createElement("li");
    periodEl.textContent = `From ${formatDate(
      period.startDate
    )} to ${formatDate(period.endDate)}`;
    pastPeriodList.appendChild(periodEl);
  });

  pastPeriodContainer.appendChild(pastPeriodHeader);
  pastPeriodContainer.appendChild(pastPeriodList);
}

function formatDate(dateString) {
  const date = new Date(dateString);
  return date.toLocaleDateString("en-US", { timeZone: "UTC" });
}

renderPastPeriods();
```

You can try the fully functioning [period tracker web app](https://mdn.github.io/pwa-examples/intro/javascript_functionality) and view the [period tracker web app source code](https://github.com/mdn/pwa-examples/tree/master/intro/javascript_functionality) on GitHub. Yes, it works, but it's not a yet PWA.

## Up next

> At it's core, a PWA is a web application that is progressively enhanced to work offline. Now that we have a fully functional web application, we add the features required to convert into a PWA, including the [manifest file](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Manifest_file), [secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure), and [service worker](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Service_worker), we have to make this application work.

Reviewer: ```suggestion
At its core, a PWA is a web application that can be installed is progressively enhanced to work offline. Now that we have a fully functional web application, we add the features required to convert it to a PWA, including the [manifest file](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Manifest_file), [secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure), and [service worker](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Service_worker).
```
